File: lc1/two-sum.go
package lc1

func TwoSum_0(nums []int, target int) []int {

	for a := 0; a < len(nums)-1; a++ {
		for b := a + 1; b < len(nums); b++ {
			if nums[a]+nums[b] == target {
				return []int{a, b}
			}
		}
	}

	return nil
}

// Second solution, uses a map to track expected values
func TwoSum(nums []int, target int) []int {

	// Key = diff, Value = Index such that diff + nums[index] = target
	diffMap := make(map[int]int)

	for index, value := range nums {
		if diffIndex, exists := diffMap[value]; exists {
			return []int{diffIndex, index}
		}

		diff := target - value
		diffMap[diff] = index
	}

	return nil
}

File: lc1/two-sum.js
function lc1(arr, target) {
    const valueIndexMap = {} // Maps the value to the index of that value

    for (let i=0; i<arr.length; i++) {
        const diff = target - arr[i];

        // see if needed value has already been seen
        if (valueIndexMap[diff] !== undefined) {
            return [valueIndexMap[diff], i]
        }

        valueIndexMap[arr[i]] = i;
    }
}



console.log(lc1([2,7,11,15], 9));

console.log(lc1([3,2,4], 6));

console.log(lc1([3,3], 6));